dll/win32/wdmaud.drv/legacy.c

```diff
@@ -1,89 +1,306 @@
 DWORD OpenCount = 0;
 
+/*
+    This is a wrapper around DeviceIoControl which provides control over
+    instantiated sound devices. It waits for I/O to complete (since an
+    instantiated sound device is opened _In_ overlapped mode, this is necessary).
+*/
+MMRESULT
+WdmAudIoControl(
+    _In_ PWDMAUD_DEVICE_INFO DeviceInfo,
+    _In_opt_ ULONG BufferSize,
+    _In_opt_ PVOID Buffer,
+    _In_ DWORD IoControlCode)
+{
+    MMRESULT Result = MMSYSERR_NOERROR;
+    OVERLAPPED Overlapped;
+    DWORD Transferred = 0;
+    BOOL IoResult;
+
+    /* Overlapped I/O is done here - this is used for waiting for completion */
+    ZeroMemory(&Overlapped, sizeof(OVERLAPPED));
+    Overlapped.hEvent = CreateEvent(NULL, FALSE, FALSE, NULL);
+
+    if ( ! Overlapped.hEvent )
+        return Win32ErrorToMmResult(GetLastError());
+
+    if (DeviceInfo->DeviceType != AUX_DEVICE_TYPE &&
+        DeviceInfo->DeviceType != MIXER_DEVICE_TYPE)
+    {
+        /* Enter critical section */
+        EnterCriticalSection(&WdmAudCriticalSection);
+    }
+
+    /* Store input data */
+    DeviceInfo->BufferSize = BufferSize;
+    DeviceInfo->Buffer = Buffer;
+
+    /* Talk to the device */
+    IoResult = DeviceIoControl(KernelHandle,
+                               IoControlCode,
+                               DeviceInfo,
+                               sizeof(WDMAUD_DEVICE_INFO),
+                               DeviceInfo,
+                               sizeof(WDMAUD_DEVICE_INFO),
+                               &Transferred,
+                               &Overlapped);
+
+    /* If failure occurs, make sure it's not just due to the overlapped I/O */
+    if ( ! IoResult )
+    {
+        if (GetLastError() == ERROR_IO_PENDING)
+        {
+            /* Wait for I/O complete */
+            WaitForSingleObject(Overlapped.hEvent, INFINITE);
+        }
+        Result = Win32ErrorToMmResult(GetLastError());
+    }
+
+
+    if (DeviceInfo->DeviceType != AUX_DEVICE_TYPE &&
+        DeviceInfo->DeviceType != MIXER_DEVICE_TYPE)
+    {
+        /* Leave critical section */
+        LeaveCriticalSection(&WdmAudCriticalSection);
+    }
+
+    /* Don't need this any more */
+    CloseHandle(Overlapped.hEvent);
+
+    DPRINT("Transferred %d bytes in Sync overlapped I/O\n", Transferred);
+
+    return Result;
+}
+
 DWORD
 WINAPI
-MixerEventThreadRoutine(
+WaveThreadRoutine(
     LPVOID Parameter)
 {
-    HANDLE WaitObjects[2];
-    DWORD dwResult;
+    PWDMAUD_DEVICE_INFO DeviceInfo = (PWDMAUD_DEVICE_INFO)Parameter;
+    PWAVEHDR_EXTENSION HeaderExtension;
+    PWAVEHDR WaveHeader;
+#ifndef USE_MMIXER_LIB
     MMRESULT Result;
-    WDMAUD_DEVICE_INFO DeviceInfo;
-    PSOUND_DEVICE_INSTANCE Instance = (PSOUND_DEVICE_INSTANCE)Parameter;
+#endif
+
+    /* Enter critical section */
+    EnterCriticalSection(&WdmAudCriticalSection);
+
+    while(TRUE)
+    {
+        DPRINT("WaveQueue %p\n", DeviceInfo->DeviceState.WaveQueue);
+        if (DeviceInfo->DeviceState.WaveQueue)
```

Comment:
You're checking this again below (L116), so it seems unnecessary here.

Reply:
Fixed, thanks. :+1: